engine/src/main/java/org/apache/apex/engine/api/plugin/DAGExecutionPlugin.java

```diff
@@ -41,67 +44,40 @@
  * </ul>
  *
  */
-@InterfaceStability.Evolving
-public interface DAGExecutionPlugin extends Plugin<DAGExecutionPluginContext>
+public interface DAGExecutionPlugin<T extends DAGExecutionPlugin.DAGExecutionPluginContext<DAGExecutionEvent.Type>> extends Plugin<T>
 {
-  enum DAGExecutionEventType implements EventType
-  {
-    HEARTBEAT_EVENT, STRAM_EVENT, COMMIT_EVENT
-  }
 
-  abstract class DAGExecutionEvent<T extends DAGExecutionEventType> extends Event.BaseEvent<T>
+  /**
+   * The context for the execution plugins.
+   *
+   * Following events are supported
+   * <ul>
+   *   <li>{@see DAGExecutionPluginContext.HEARTBEAT} The heartbeat from a container is delivered to the plugin after it has been handled by stram</li>
+   *   <li>{@see DAGExecutionPluginContext.STRAM_EVENT} All the Stram event generated in Stram will be delivered to the plugin</li>
+   *   <li>{@see DAGExecutionPluginContext.COMMIT_EVENT} When committedWindowId changes in the platform an event will be delivered to the plugin</li>
+   * </ul>
+   *
+   */
+  interface DAGExecutionPluginContext<T extends DAGExecutionEvent.Type> extends PluginContext<T, DAGExecutionEvent<T>>
```

Comment:
Isn't it limiting to fix DAGExectionEvent<T> in PluginContext<T, DAGExecutionEvent<T>> 
vs 
DAGExecutionPluginContext<T extends DAGExecutionEvent.Type, E extends DAGExecutionEvent<T>>

On further thought this overall declaration might become cumbersome to the user like below
DAGExecutionPlugin<DAGExecutionPlugin.DAGExecutionPluginContext<DAGExecutionEvent.Type>
If the types of event type and event are removed from class and attached to method instead it might become simpler.